src/ewmh.h

```diff
@@ -69,51 +67,66 @@ enum {
 
+typedef enum {
+    // see icccm:
+    // http://www.x.org/releases/X11R7.7/doc/xorg-docs/icccm/icccm.html#WM_STATE_Property
+    WmStateWithdrawnState = 0,
+    WmStateNormalState    = 1,
+    WmStateIconicState    = 3,
+} WmState;
+
 class HSTag;
 class Client;
+class Root;
+class XConnection;
 
 extern Atom g_netatom[NetCOUNT];
 
 extern const std::array<const char*,NetCOUNT> g_netatom_names;
 
-void ewmh_init();
-void ewmh_destroy();
-void ewmh_update_all();
+class Ewmh {
+public:
+    Ewmh(XConnection& xconnection);
+    ~Ewmh();
 
-void ewmh_add_client(Window win);
-void ewmh_remove_client(Window win);
-void ewmh_set_wmname(const char* name);
-void ewmh_update_wmname();
+    void injectDependencies(Root* root);
+    void updateAll();
 
-void ewmh_update_client_list();
-void ewmh_get_original_client_list(Window** buf, unsigned long *count);
-void ewmh_update_client_list_stacking();
-void ewmh_update_desktops();
-void ewmh_update_desktop_names();
-void ewmh_update_active_window(Window win);
-void ewmh_update_current_desktop();
-void ewmh_update_window_state(Client* client);
-void ewmh_update_frame_extents(Window win, int left, int right, int top, int bottom);
-bool ewmh_is_window_state_set(Window win, Atom hint);
-bool ewmh_is_fullscreen_set(Window win);
-void ewmh_clear_client_properties(Window win);
+    void addClient(Window win);
+    void removeClient(Window win);
+    void setWmName(const char* name);
+    void updateWmName();
 
-// set the desktop property of a window
-void ewmh_window_update_tag(Window win, HSTag* tag);
+    void updateClientList();
+    void getOriginalClientList(Window** buf, unsigned long *count);
+    void updateClientListStacking();
+    void updateDesktops();
+    void updateDesktopNames();
+    void updateActiveWindow(Window win);
+    void updateCurrentDesktop();
+    void updateWindowState(Client* client);
+    void updateFrameExtents(Window win, int left, int right, int top, int bottom);
+    bool isWindowStateSet(Window win, Atom hint);
+    bool isFullscreenSet(Window win);
+    void clearClientProperties(Window win);
 
-void ewmh_handle_client_message(Root* root, XEvent* event);
+    // set the desktop property of a window
+    void windowUpdateTag(Window win, HSTag* tag);
 
-void ewmh_set_window_opacity(Window win, double opacity);
+    void handleClientMessage(XEvent* event);
 
-typedef enum {
-    // see icccm:
-    // http://www.x.org/releases/X11R7.7/doc/xorg-docs/icccm/icccm.html#WM_STATE_Property
-    WmStateWithdrawnState = 0,
-    WmStateNormalState    = 1,
-    WmStateIconicState    = 3,
-} WmState;
+    void setWindowOpacity(Window win, double opacity);
+
+    void windowUpdateWmState(Window win, WmState state);
 
-void window_update_wm_state(Window win, WmState state);
+    static Ewmh& get(); // temporary singleton getter
+
+private:
+    bool focusStealingAllowed(long source);
+    bool readClientList(Window** buf, unsigned long *count);
+    Root* root_ = nullptr;
+    XConnection& X;
```

Comment:
This should be a lower-case `x`, shouldn't it?

Reply:
Right now, I always use `X` in classes where `X` is actually used, because I think `X.display()` looks much better than `x.display()` or `x_.display()` or `X_.display()`. A lower-case `x` looks like the `x`-coordinate.

Reply:
Mhm, I see the issue. How do you think about `xConn`?

Reply:
I think `xConn.display()` in every second line looks neither helpful nor nice. What's the issue with `X` (or maybe with `X_`)? Is this only about obeying the law "Every variable name must be in camelCase" or is there any risk of confusion where the variable is used? As a compromise, we can name it `X_` to make it clear that `X_` is a member variable.

Reply:
It's really only about staying with the naming convention, so it's truly a minor concern. I can live with `X`/`X_`, but I wanted to check if there's maybe a third way that makes everybody happy.

Adding the underscore suffix would be appropriate regardless of the casing, but since we're breaking the rules anyway, just go with what you prefer here. ;-)

Reply:
I've named it `X_` to make it easier for the reader to see that this is a member variable, even though it looks a bit different because of the upper case X.